pkg/utils/k8sutil.go

```diff
@@ -0,0 +1,117 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/keikoproj/addon-manager/pkg/client/informers/externalversions/internalinterfaces"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+)
+
+// GetClient returns a k8s clientset to the request from inside of cluster
+func GetClient() kubernetes.Interface {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		logrus.Fatalf("Can not get kubernetes config: %v", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		logrus.Fatalf("Can not create kubernetes client: %v", err)
+	}
+
+	return clientset
+}
+
+func BuildOutOfClusterConfig() (*rest.Config, error) {
+	kubeconfigPath := os.Getenv("KUBECONFIG")
+	if kubeconfigPath == "" {
+		return nil, fmt.Errorf("did not find kubeconfig")
+	}
+	return clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+}
+
+// GetClientOutOfCluster returns a k8s clientset to the request from outside of cluster
+func GetClientOutOfCluster() kubernetes.Interface {
+	config, err := BuildOutOfClusterConfig()
+	if err != nil {
+		logrus.Fatalf("Can not get kubernetes config: %v", err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil
+	}
+	return clientset
+}
+
+func NewWorkflowInformer(dclient dynamic.Interface, ns string, resyncPeriod time.Duration, indexers cache.Indexers, tweakListOptions internalinterfaces.TweakListOptionsFunc) cache.SharedIndexInformer {
+	resource := schema.GroupVersionResource{
+		Group:    "argoproj.io",
+		Version:  "v1alpha1",
+		Resource: "workflows",
+	}
+	informer := NewFilteredUnstructuredInformer(
+		resource,
+		dclient,
+		ns,
+		resyncPeriod,
+		indexers,
+		tweakListOptions,
+	)
+	return informer
+}
+
+func NewFilteredUnstructuredInformer(resource schema.GroupVersionResource, client dynamic.Interface, namespace string, resyncPeriod time.Duration, indexers cache.Indexers, tweakListOptions internalinterfaces.TweakListOptionsFunc) cache.SharedIndexInformer {
```

Comment:
Same, why are we switching from typed to dynamic clients?

Reply:
workflow informer has always been dynamic https://github.com/keikoproj/addon-manager/blob/master/controllers/controller.go#L10. 
Just switch to cached.ShareIndexInformer for efficiency.

Reply:
db6f893de3812e442ece0edc1be00e9732a6902d